cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -342,6 +344,53 @@ class GcsFileSystem::Impl {
     return internal::ToArrowStatus(CreateDirMarkerRecursive(p.bucket, p.object));
   }
 
+  Status DeleteDir(const GcsPath& p, const io::IOContext& io_context) {
+    RETURN_NOT_OK(DeleteDirContents(p, io_context));
+    if (!p.object.empty()) {
+      return internal::ToArrowStatus(client_.DeleteObject(p.bucket, p.object));
+    }
+    return internal::ToArrowStatus(client_.DeleteBucket(p.bucket));
+  }
+
+  Status DeleteDirContents(const GcsPath& p, const io::IOContext& io_context) {
+    // Deleting large directories can be fairly slow, we need to parallelize the
+    // operation. This uses `std::async()` to run multiple delete operations in parallel.
+    // A simple form of flow control limits the number of operations running in
+    // parallel.
+
+    auto async_delete =
+        [&p](gcs::Client& client,
+             google::cloud::StatusOr<gcs::ObjectMetadata> o) -> google::cloud::Status {
```

Comment:
Two suggestions:
1) just capture the client as a closure parameter
2) return an arrow `Status` here, such that you can use `AllFinished(const std::vector<Future>&)`

Reply:
Done.  I captured `this` because capturing `client_` is not possible in C++11.